File: src/types/api-types.ts
export enum FileIndexingStatusEnum {
	FAILED = "FAILED",
	FINISHED = "FINISHED",
	INDEXING = "INDEXING",
}

export enum UserRoleEnum {
	ADMIN = "ADMIN",
	MEMBER = "MEMBER",
	OWNER = "OWNER",
}

export interface APIError {
	details?: string;
	message: string;
}

export interface ApplicationDetails {
	background_context?: string;
	hypothesis?: string;
	impact?: string;
	milestones_and_timeline?: string;
	novelty_and_innovation?: string;
	preliminary_data?: string;
	rationale?: string;
	research_feasibility?: string;
	risks_and_mitigations?: string;
	scientific_infrastructure?: string;
	team_excellence?: string;
}

export interface ApplicationDraftCompleteResponse {
	id: string;
	status: "complete";
	text: string;
}

export interface ApplicationDraftProcessingResponse {
	id: string;
	status: "generating";
}

export type ApplicationDraftResponse = ApplicationDraftCompleteResponse | ApplicationDraftProcessingResponse;

export interface CreateApplicationRequestBody {
	cfp_url?: string;
	title: string;
}

export interface CreateOrganizationRequestBody {
	abbreviation?: null | string;
	full_name: string;
}

// Workspace Types
export interface CreateWorkspaceRequestBody {
	description?: null | string;
	logo_url?: null | string;
	name: string;
}

// Organization Types
export interface FundingOrganization extends TableIdResponse {
	abbreviation: null | string;
	full_name: string;
}

export interface FundingOrganizationBase extends TableIdResponse {
	abbreviation: null | string;
	full_name: string;
}

export interface GrantApplication extends TableIdResponse {
	completed_at: null | string;
	details: ApplicationDetails | null;
	grant_template: GrantTemplate | null;
	research_objectives: null | ResearchObjective[];
	text_generation_results: null | TextGenerationResult[];
	title: string;
	workspace_id: string;
}

export interface GrantApplicationFile {
	grant_application_id: string;
	rag_file: RagFileResponse;
}

export interface GrantTemplate extends TableIdResponse {
	funding_organization: FundingOrganizationBase | null;
	grant_sections: GrantSection[];
	name: string;
	template: string;
}

export interface LoginRequestBody {
	id_token: string;
}

export interface LoginResponse {
	jwt_token: string;
}

export interface OrganizationFile {
	funding_organization_id: string;
	rag_file: RagFileResponse;
}

// Auth Types
export interface OTPResponse {
	otp: string;
}

// File Types
export interface RagFileResponse extends TableIdResponse {
	filename: string;
	indexing_status: FileIndexingStatusEnum;
	mime_type: string;
	size: number;
}

export interface ResearchObjective {
	description?: string;
	number: number;
	relationships?: string[];
	research_tasks: ResearchTask[];
	title: string;
}

// Application Types
export interface ResearchTask {
	description?: string;
	number: number;
	relationships?: string[];
	title: string;
}

export interface TableIdResponse {
	id: string;
}

export interface TextGenerationResult {
	content: string;
	type: string;
}

export interface UpdateApplicationRequestBody {
	research_objectives?: ResearchObjective[];
	title?: string;
}

export interface UpdateOrganizationRequestBody {
	abbreviation?: null | string;
	full_name?: string;
}

export interface UpdateWorkspaceRequestBody {
	description?: null | string;
	logo_url?: null | string;
	name?: string;
}

export interface Workspace extends WorkspaceBaseResponse {
	grant_applications: GrantApplication[];
}

export interface WorkspaceBaseResponse extends TableIdResponse {
	description: null | string;
	logo_url: null | string;
	name: string;
	role: UserRoleEnum;
}

File: src/api_types.py
from types import SimpleNamespace
from typing import Any, Literal, NotRequired, TypedDict

from sanic import Request, Sanic
from sqlalchemy.ext.asyncio import async_sessionmaker

from src.db.enums import UserRoleEnum
from src.db.json_objects import ResearchObjective


class RequestContext(SimpleNamespace):
    """The context of an API request."""

    firebase_uid: str
    """The Firebase User ID."""

    session_maker: async_sessionmaker[Any]
    """The session maker."""


APIRequest = Request[Sanic[Any, RequestContext], RequestContext]


class TableIdResponse(TypedDict):
    """A base response containing only a row ID."""

    id: str
    """The ID of the application."""


# Organization API Types
class CreateOrganizationRequestBody(TypedDict):
    """The request body for creating a funding organization."""

    full_name: str
    abbreviation: str | None


class UpdateOrganizationRequestBody(TypedDict):
    """The request body for updating a funding organization."""

    full_name: NotRequired[str]
    abbreviation: NotRequired[str | None]


# Workspace API Types
class CreateWorkspaceRequestBody(TypedDict):
    """The request body for creating a workspace."""

    name: str
    """The name of the workspace."""
    description: str | None
    """The description of the workspace."""
    logo_url: NotRequired[str | None]
    """The URL of the workspace logo."""


class UpdateWorkspaceRequestBody(TypedDict):
    """The request body for updating a workspace."""

    name: NotRequired[str]
    """The name of the workspace."""
    description: NotRequired[str | None]
    """The description of the workspace."""
    logo_url: NotRequired[str | None]
    """The URL of the workspace logo."""


class WorkspaceBaseResponse(TableIdResponse):
    """Base response for retrieving workspaces."""

    name: str
    """The name of the workspace."""
    description: str | None
    """The description of the workspace."""
    logo_url: str | None
    """The URL of the workspace logo."""
    role: UserRoleEnum
    """The role of the user in the workspace."""


# Application API Types
class CreateApplicationRequestBody(TypedDict):
    """The request body for creating an application."""

    title: str
    """The title of the application."""
    cfp_url: NotRequired[str]
    """Grant CFP URL."""


class UpdateApplicationRequestBody(TypedDict):
    """The request body for updating an application."""

    title: NotRequired[str]
    """The title of the application."""
    research_objectives: NotRequired[list[ResearchObjective]]
    """The research objectives of the application."""


class ApplicationDraftProcessingResponse(TypedDict):
    """The response schema for an application draft in processing state."""

    id: str
    """The ID of the grant application draft."""
    status: Literal["generating"]
    """The status of the grant application draft."""


class ApplicationDraftCompleteResponse(TypedDict):
    """The response schema for a completed application draft."""

    id: str
    """The ID of the grant application draft."""
    status: Literal["complete"]
    """The status of the grant application draft."""
    text: str


# Auth API Types
class OTPResponse(TypedDict):
    """The response body for the OTP endpoint."""

    otp: str
    """The otp identifying the user."""


class LoginRequestBody(TypedDict):
    """The request body for the login endpoint."""

    id_token: str
    """The ID token from Firebase."""


class LoginResponse(TypedDict):
    """The response body for the login endpoint."""

    jwt_token: str
    """The JWT token identifying the user."""
